### DISCUSSION TITLE
Multiple smithy services?

### DISCUSSION THREAD

**Discussion by Quafadas** [disneystreaming/smithy4s#238] [Q&A] [answered]
I'm experimenting, so apologies if this question is naive. I'm expanding on the website examples, in this case to have 2 smithy services. In addition to the "hello" service, I've written out a simple ```TodoService.smithy```.

Now, I'd like to turn it into an http4s server.
```
object Main extends IOApp {
    
    def run(args: List[String]): IO[ExitCode] = 

        val helloRoutes = SimpleRestJsonBuilder.routes(HelloWorldImpl).make
        val todoRoutes = SimpleRestJsonBuilder.routes(TodoImpl).make
        val docHelloRoutes = docs[IO](HelloWorldService)
        val docTodoRoutes = docs[IO](TodoService)

        val allRoutes = (helloRoutes <+> todoRoutes ).map(in => 
                in <+> docHelloRoutes <+> docTodoRoutes
            ).map(_.orNotFound)

        val server = allRoutes.map {
            routes => 
                EmberServerBuilder.default[IO]
                .withPort(port"8080")
                .withHost(host"localhost")
                .withHttpApp(routes)
                .build
        }.map(_.use(_ => IO.never).as(ExitCode.Success))
        
        server.getOrElse(IO(ExitCode.Error))       
}
```
This compiles and runs, but has a couple of fairly severe downsides. 

1. The Todo routes, don't actually work! I would have expected it to, based on what I've seen of http4s. To be clear, I'm have very limited experience with http4s, so it may be that my attempt is "wrong".
2. The docs default to the "hello" swagger pages. The "Todo" swagger does exist, but one has to know about the existence of the service, in order to find it. I can't see a hint on the swagger UI, on how to navigate to the second service.

So the key question would be whether I am doing something am I doing something that is obviously "wrong" trying to "combine" these routes / docs in this manner? 

Is the intent to have "one smithy service per http server"? This looks like the obvious solution to all the above :-)... but I want to make sure I'm not doing something silly?

The full example is here, in case the question is poorly formed. 
https://github.com/Quafadas/mill-full-stack

**Comment by Baccata** [comment, marked as answer]
Hey there, those are pretty good question. Also, thanks for accompanying them with a repo, it greatly helps formulate an answer. 

#### 1. Why are the todo routes not working ? 

It's because your `<+>` call is working at the level of `Either`, not at the level of `HttpRoute`. Therefore, it returns the first non-left value, as oppose to an either containing the sum of both successes. 

See the [instance for `SemigroupK[Either[E, *]]](https://github.com/typelevel/cats/blob/708cbaba2e6ac07a5ac1819b6c18b0c83a2faaba/core/src/main/scala/cats/instances/either.scala#L210-L223)

So your [allRoutes](https://github.com/Quafadas/mill-full-stack/blob/master/backend/src/example/Server.scala#L21-L23) value should look like : 

```scala 
val allRoutes = (helloRoutes, todoRoutes).mapN(_ <+> _ ).map(_ <+> docHelloRoutes <+> docTodoRoutes).map(_.orNotFound)
```

#### 2. Why do the docs UX suck when there's more than one service

Well, we're very aware of this. There's an [open issue](https://github.com/disneystreaming/smithy4s/issues/150) to make it better, but also we're currently facing a bigger problem that the latest swagger-ui doesn't work anymore with how we use it. See [this release](https://github.com/swagger-api/swagger-ui/releases/tag/v4.1.3). So the whole swagger-ui is bound to receive an overhaul in the near future, and when it does we'll kill two birds in one stone and make it work with several service definitions.

**Reply by Quafadas** [reply]
> It's because your <+> call is working at the level of Either

Arg. Thanks for this - not related to smithy at all... I possibly should have got this one on my own. Great answer, much appreciated.

>  So the whole swagger-ui is bound to receive an overhaul in the near future, and when it does we'll kill two birds in one stone and make it work with several service definitions.

Sounds like you've had the rug pulled out from under you a little on this one - I hope it's not too painful! Thanks for the response.
